### DISCUSSION TITLE
Improve the State Management

### DISCUSSION THREAD

**Discussion by aaschmitz** [melonjs/melonJS#1044] [Ideas]
The idea is perform non-destructive State swaps, for use in situations like "Inventory Screen" or "Pause Screen" (accessed by "Play Screen" without destroying the same), as the article below:

http://caspervonb.github.io/2014/03/02/javascript-game-development-game-states.html

"When pushing a state, the previous state will be paused and the new state will be initialized, when popping a state the state is disposed and the previous state is resumed.

This will allow us to temporarily push new states on top on top of the stack, without destroying any of the data of the previous states."

**Comment by parasyte** [comment]
Interesting concept. We have some similar capabilities with `me.ObjectContainer`, but for sure it needs to be manageable within a stack context. There was some old discussion about doing this, too: https://github.com/melonjs/melonJS/issues/59#issuecomment-14351524

My proposal is to keep the current state machine (which supports changing states, and within the current state: pause-resume and stop-start) and build this new functionality as a "modal stack" feature. Modality can be simulated (sort of) using the pause-resume feature in a state, but it can only ever be one level deep (pause is a boolean). As a stack, the modality can be multiple levels deep, which is important for deeply nested menu interfaces and so on.

For example, it's probably enough to just swap the `me.game.world` reference with a new `me.ObjectContainer` ... And tada! You've got a brand new empty world to add whatever you like. When you're done, replace `me.game.world` with the original container reference!

If this flow is made easier with a simple API, then you get the same kind of benefits as described in the article. Every entity in the container needs to support new callbacks, described in the article as `pause` and `resume`. But those names are already in use by the state machine. We might name them something like `onModalPause` and `onModalResume`. (These callbacks are necessary for properly managing event handlers.)

The final point to cover is compositing different modal containers, E.g. a pause modal probably still wants to draw the level and entities below the pause menu. All it needs for this is a reference to its "parent" container (the original `me.game.world`) and call its draw method.

As for API? I guess we need a new `me.game.modal` module that has methods:
- `function push(container, callback)` - Returns previously active container
- `function pop(callback)` - Returns newly active container

The callbacks are optional, and they accept an argument which references the container that was pushed/popped. Both functions are asynchronous, because the actual `me.game.world` switch needs to be deferred. (See #455; the same principle applies here.)

---

Example: Simple pause menu

``` javascript
// Create a pause menu
var container = new me.ObjectContainer();
var renderer = new game.Renderer();
container.addChild(renderer);
container.addChild(new me.ColorLayer("background", "rgba(0, 0, 0, 0.75)", 1));
container.addChild(new me.ImageLayer("menu_bg", 300, 500, "menu_bg", 2));
container.addChild(new game.Button(100, 50, "Options"));
container.addChild(new game.Button(100, 100, "Restart"));
container.addChild(new game.Button(100, 150, "Exit"));

// Start pause menu modal state
renderer.modalParent = me.game.modal.push(container);
```

The hypothetical `game.Renderer` class only calls `this.modalParent.draw()` within its draw method.

---

Example: Load a new sub-level

``` javascript
me.game.modal.push(new me.ObjectContainer(), function (container) {
    // This callback fires when the new modal state is ready
    // `container` is a reference to the pushed ObjectContainer
    me.levelDirector.loadLevel("sublevel");
});
```

Example: Return to previous level with all entity state preserved:

``` javascript
me.game.modal.pop(function (container) {
    // This callback fires when the previous modal state is ready
    // `container` is a reference to the popped ObjectContainer
    console.log("Welcome back!");
});
```
